### DISCUSSION TITLE
Two-stage tokenization

### DISCUSSION THREAD

**Discussion by nnmm** [squidfunk/mkdocs-material#4276] [Ideas]
We are experiencing not-so-great search results. I believe the root cause is that compound words like "CamelCase" are *not* in the search index, since they are split by the tokenizer separator, in this example into "Camel" and "Case".

I want to propose keeping both the subdivided word and the original word in the index.

This could be achieved by tokenizing first into "words", i.e. splitting on whitespace + punctuation, but not case changes, and adding those words to the token list. Afterwards, these words can be "subdivided" with another regex, which would do case changes, and if any subdividing happened, the parts are added to the token list too. 

On the query side, we would apply the "word tokenization", but not necessarily the "subdivision tokenization".

I'm using case change splitting as an example here, but in our case, it's also important to move the splitting on "::" to the "subdivision tokenization" stage to make C++ API docs much more searchable.

Please let me know what you think! Unfortunately I'm not a TypeScript dev or I'd open a PR myself.

**Comment by squidfunk** [comment]
Thanks for reporting. As noted in my email in our previous conversation, I'm going to need a _minimal_ reproduction that showcases the issue. You can attach it here as a zip file. It's likely fixable with a slightly different configuration, but I need to know exactly what you're doing in order to help you.

**Reply by nnmm** [reply]
Hi @squidfunk, it took me a little while, but here is a minimal example. Search for "BoundingBox" in this demo, and you'll find that the `BoundingNotBox` article is ranked above¹ `BoundingBox`. The desired result would be to not find `BoundingNotBox` at all, or at least with a much lower score. You also will not get any results for "boundingbox".

[minimal_example.zip](https://github.com/squidfunk/mkdocs-material/files/9469627/minimal_example.zip)

¹ I added a few "bounding"  and "box" to the body of the `BoundingBox` article to ensure this – I believe without this, they would be tied.

**Reply by squidfunk** [reply]
Thanks for providing the example. Here's what I'm seeing:

<img width="686" alt="Bildschirmfoto 2022-09-01 um 15 12 29" src="https://user-images.githubusercontent.com/932156/187922355-f704bc9a-9edb-4c0c-9fbf-ff4355a3671f.png">

Two observations:

1. The __BoundingNotBox__ article is ranked higher because there are more matches of both words or stems of words in the document. This is the behavior of lunr.js. If I remove line 7 from `bounding_not_box.md`, the pages are swapped which is what you'd expect. You can inspect the computed ranks for pages by looking at the HTML of the search results and checking `data-md-score`:

    <img width="669" alt="Bildschirmfoto 2022-09-01 um 15 13 29" src="https://user-images.githubusercontent.com/932156/187922581-c61cdf74-69ad-4a2a-8e37-cceb79c4f08e.png">

    In theory, we could expose the [two parameters](https://lunrjs.com/guides/customising.html#similarity-tuning) of lunr.js with which scoring can be tuned as configuration options.

2. Searching for __boundingbox__ does not yield any results because tokenization split the documents into two words, namely __Bounding__ and __Box__. The same is done for the query. Thus, if you search for __boundingbox__, there's nothing to split, and thus no results are found. I guess this is why you're asking for two-step tokenization, as you'd want to add both variants to the index, but that's currently not possible and I'm not sure how it should work, because if we'd do that for all words, it would blow up the index size.

**Reply by squidfunk** [reply]
So as a conclusion: I'm happy to look into improving search and adding more functionality for additional use cases, but we'd need a clear problem description of what doesn't work and what we'd need to make it work. Without a goal, it's hard to provide further mechanisms for tuning search results. Please understand that as all features, the search must work for all cases and all languages, so it's natural that it's not a 100% solution for everyone, but it should yield good results in most cases. Your case is clearly something that could be improved, but we'll need to check whether the efforts and cost (implementation, maintenance and runtime) are worth.

**Reply by nnmm** [reply]
Thanks for the quick reply, and for pointing me to the `data-md-score`. By "cheating" a bit with the extra terms in line 7, I only wanted to demonstrate that the scores must be pretty close, and it's easily possible for the wrong article to overtake the correct article. The `data-md-score` shows that more directly – here are the scores again with line 7 removed: 

![bounding_box_ranking](https://user-images.githubusercontent.com/1933253/187941307-7813ccbe-4644-47cd-b0a1-8765f24d4f2f.png)

So, as I'm sure you already understand, `BoundingNotBox` looks exactly the same as `BoundingBox` to the search plugin when queried for `BoundingBox`. That's a pretty fundamental limitation, which is why I think score-tweaking approaches would not be as clean of a solution as the two-stage tokenization.

<hr>



>  Searching for **boundingbox** does not yield any results because tokenization split the documents into two words, namely **Bounding** and **Box**. The same is done for the query. Thus, if you search for **boundingbox**, there's nothing to split, and thus no results are found. I guess this is why you're asking for two-step tokenization, as you'd want to add both variants to the index

Yes, that's one thing we would gain, the other would be more robust ranking.

> , but that's currently not possible and I'm not sure how it should work, because if we'd do that for all words, it would blow up the index size.

Would you be convinced if I implemented the feature and tried to measure the effect on the index size? Theoretically it seems to me like it should only make a small difference, maybe 10% or 15% for code-heavy docs like ours.

**Reply by nnmm** [reply]
> So as a conclusion: I'm happy to look into improving search and adding more functionality for additional use cases, but we'd need a clear problem description of what doesn't work and what we'd need to make it work. Without a goal, it's hard to provide further mechanisms for tuning search results. Please understand that as all features, the search must work for all cases and all languages, so it's natural that it's not a 100% solution for everyone, but it should yield good results in most cases. Your case is clearly something that could be improved, but we'll need to check whether the efforts and cost (implementation, maintenance and runtime) are worth.

I think this can be done in a pretty simple way. Before we had

```
separator: /*split on whitespace, punctuation, case changes, dashes etc.*/
```

and afterwards we'd have
```
separator: /*split on whitespace, punctuation */
extra_search_term_separator: /*split on case changes, dashes etc.*/
```

where `extra_search_term_separator` is applied in the `tokenizer()` function on top of the `separator`. I'm happy to measure runtime difference of that. Queries are only split by `separator`, like now. That part would get faster, if anything, since the regex would be smaller.

You can easily see that this is equivalent to the current mechanism when `extra_search_term_separator` is empty. So from my (perhaps naive) point of view, this solution would be strictly better than the current implementation, apart from a potential performance penalty.

**Reply by squidfunk** [reply]
> So, as I'm sure you already understand, BoundingNotBox looks exactly the same as BoundingBox to the search plugin when queried for BoundingBox. That's a pretty fundamental limitation, which is why I think score-tweaking approaches would not be as clean of a solution as the two-stage tokenization.

Yes. This is a limitation of lunr.js. lunr.js doesn't take token positions into account when scoring, which would probably give better results. Yet, I haven't found another suitable replacement with the same support for the 58 languages we support.

I doubt that adding more steps for configuring and tweaking search would lead to more robust rankings. In fact, putting more steps into the pipeline is likely to degrade into something that's quite impossible to debug – it's already a challenge.

> Would you be convinced if I implemented the feature and tried to measure the effect on the index size? Theoretically it seems to me like it should only make a small difference, maybe 10% or 15% for code-heavy docs like ours.

I'd be convinced if the feature (or say enhancement) ticks all the boxes which need to be ticked before something can go into this project (recommended read: #2399) and other projects are not badly impacted, i.e. it's opt-in. With all that given I still can't promise that it's going into `master`, but it's definitely a good experiment for sure, and maybe we can work together and find a good solution. In the end, it's always me who has to maintain this project, so please understand that I'm extremely careful about what I commit to maintain. After all, having a limited scope and many components that work well together is what makes this project successful.

**Reply by squidfunk** [reply]
BTW, as it's not apparent from this conversation – it's related to the Insiders search, as the search in the community edition doesn't allow for advanced tokenization. Thus, a potential PR should be submitted on the Insiders repository.

**Reply by squidfunk** [reply]
> where extra_search_term_separator is applied in the tokenizer() function on top of the separator. I'm happy to measure runtime difference of that. Queries are only split by separator, like now. That part would get faster, if anything, since the regex would be smaller.

The runtime differences for tokenizing the query are neglible – queries are usually shorter than a few dozen characters. In any case, measuring is a good idea. Note that you'll need to make sure not to mess up search highlighting and ranking, as you're now adding many terms to the index twice.

**Reply by flynneva** [reply]
@squidfunk just a follow-up question here but does the current search implementation add the _raw search string_ to the token list? To me this would be a super simple work around that would get you almost to the approach that @nnmm mentions above.  Essentially the search assumes the raw input string is exactly what the user is searching for (which IMO should be done most of the time).

EDIT: I guess this could easily be done today just by adding to the separator regex, no?  Where the first regex group just matches everything...without splitting the raw string at all. EDIT2: nope this does not work

**Reply by squidfunk** [reply]
@flynneva sorry, but I'm not sure I understand what you mean. Could you maybe provide an example? What do you mean by token list?

**Reply by flynneva** [reply]
@squidfunk you can ignore my question above - after digging into how the search is implemented I realized it didn't make sense 😅 sorry!

**Comment by nnmm** [comment]
For the record, here is a JS implementation for an arbitrary number of stages using recursion.
I tried it on our docs and it gives great results.

```js
function fine_split(inputs, separators) {
  if (separators.length === 0) {
    return inputs
  }
  let [separator, ...remaining_separators] = separators
  return inputs.flatMap(tok => {
    let tokens = tok.split(separator)
    const add_back_original_token = tokens.length > 1
    tokens = fine_split(tokens, remaining_separators)
    if (add_back_original_token) {
      tokens.push(tok)
    }
    return tokens.filter(tok => tok.length > 0)
  })
}

function tokenize(input) {
  const inputs = Array.isArray(input) ? input : [input]
  let tokens = inputs.flatMap(inp => extract(inp));

  // hierarchical_separators is a list of regexes
  let [primary_separator, ...fine_separators] = hierarchical_separators
  tokens = tokens.flatMap(tok => tok.split(primary_separator))
  tokens = fine_split(tokens, fine_separators)
  return tokens
}
```
